refactor(traverse): enter node before entering scope

overlookmotel opened the pull request:
Closes #4200.

Align `Traverse`'s behavior with `Visit` and `VisitMut`. For types with scopes, call `enter_*` before entering scope, and call `exit_*` after exiting scope.

overlookmotel commented:
* **#4685** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4685?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#4684** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4684?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Actually, this needs further changes. Tests aren't picking it up, but it breaks a few things.

overlookmotel commented:
Fixed now and ready for review.

Boshen commented:
I removed the `!` minor version bump because this is a bug fix and not user facing breaking change.

overlookmotel committed:
refactor(traverse): enter node before entering scope (#4684)

Closes #4200.

Align `Traverse`'s behavior with `Visit` and `VisitMut`. For types with scopes, call `enter_*` before entering scope, and call `exit_*` after exiting scope.
